larg4/Services/LArG4Detector_service.cc

```diff
@@ -321,24 +321,34 @@ void larg4::LArG4DetectorService::setStepLimits(G4GDMLParser *parser){
                   << " appropriate parameter.";
 
   std::string volumeName  = "";
-  float previousStepLimit = 0.;
-  float newStepLimit      = 0.;
-  for(size_t i=0; i<inputVolumes_; ++i)
+  G4LogicalVolume* setVol = 0;
+  for(auto const& [name, newStepLimit] : overrideGDMLStepLimit_Map )
   {
+    G4double previousStepLimit = 0.;
+
     // -- Check whether the volumeName provided corresponds to a valid volumeName in the geometry
-    G4LogicalVolume* setVol = parser->GetVolume(selectedVolumes_[i].first);
+    try {
+      setVol = volStore->GetVolume(name);
+    } catch (...)
+    {
+      throw cet::exception("invalidInputVolumeName") << "Provided volume name : "
+                                                     << name << " not found!\n";
+      continue;
+    }
 
     // -- get the G4LogicalVolume corresponding to the selectedVolume
     volumeName = setVol->GetName();
-    newStepLimit = selectedVolumes_[i].second;
     MF_LOG_DEBUG("LArG4DetectorService::setStepLimits") << "Got logical volume with name: "
                                                         << volumeName;
 
+    G4UserLimits* fStepLimitOverride = new G4UserLimits();
+    G4AutoDelete::Register(fStepLimitOverride);
```

Comment:
In the examples of setting user limits that I could find, I always saw them declared dynamically that way. The examples didn't actually delete them; however, valgrind does complain about them if you don't.

I tried your suggestion, but I get segfaults when I try to run it. This is the same thing that happened when I tried to use **new** and **delete**. I'm not sure if it's because the LogicalVolumes are themselves internally registered when added to the LogicalVolumeStore or if something else is preventing the user limits from going out of scope once set for the volume.

Making it a stack variable does seems to work fine in the first loop (subject of your next comment), but previously, I didn't have any issues in that loop calling `delete fStepLimit;` -- only within my `setStepLimits()` function. I tried to move the functionality of my function to that first block to see if that would work, but then it was no longer OK with me calling delete. I did some desperate googling and I chanced upon the G4AutoDelete class and that seemed to work. From what I can tell, this is most commonly used for automatic garbage collection in a thread safe environment. The G4 manual does says that G4LogicalVolume is a **thread-shared** class (at least for the 10.6 release, see page 46 of [Book for toolkit developers](http://geant4-userdoc.web.cern.ch/geant4-userdoc/UsersGuides/ForToolkitDeveloper/fo/BookForToolkitDevelopers.pdf) .

Reply:
Hmmm...then I wouldn't bother adopting my suggestion (here or in the other place) if it is this complicated.  G4 (and also ROOT) tends to self-register things, and that results in nasty ownership issues.  Feel free to ignore this comment and the other one.  But thanks for trying!